src/org/zaproxy/zap/extension/replacer/ExtensionReplacer.java

```diff
@@ -0,0 +1,263 @@
+    }
+
+    @Override
+    public void unload() {
+        super.unload();
+        // TODO Change to call HttpSender.removeListener(this) once its available in the zap.jar
+    }
+
+    private OptionsReplacerPanel getOptionsReplacerPanel() {
+        if (optionsReplacerPanel == null) {
+            optionsReplacerPanel = new OptionsReplacerPanel();
+        }
+        return optionsReplacerPanel;
+    }
+
+    protected ReplacerParam getParams() {
+        if (params == null) {
+            params = new ReplacerParam();
+        }
+        return params;
+    }
+
+    @Override
+    public String getAuthor() {
+        return Constant.ZAP_TEAM;
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(PREFIX + ".desc");
+    }
+
+    @Override
+    public URL getURL() {
+        try {
+            return new URL(Constant.ZAP_EXTENSIONS_PAGE);
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public int getListenerOrder() {
+        return 0;
+    }
+
+    private ZapMenuItem getReplacerMenuItem() {
+        if (replacerMenuItem == null) {
+            replacerMenuItem = new ZapMenuItem(PREFIX + ".topmenu.tools.shortcut", 
+                    KeyStroke.getKeyStroke(KeyEvent.VK_R, Toolkit.getDefaultToolkit().getMenuShortcutKeyMask(), false));
+
+            replacerMenuItem.addActionListener(new java.awt.event.ActionListener() {
+                @Override
+                public void actionPerformed(java.awt.event.ActionEvent ae) {
+                    Control.getSingleton().getMenuToolsControl().options(OptionsReplacerPanel.PANEL_NAME);
+                }
+            });
+        }
+        return replacerMenuItem;
+    }
+
+    private boolean contains (String original, String match, Pattern p) {
+        if (p != null) {
+            return p.matcher(original).find();
+        } else {
+            return original.contains(match);
+        }
+    }
+
+    private String replace (String original, String match, Pattern p, String replacement) {
+        if (p != null) {
+            return p.matcher(original).replaceAll(replacement);
+        } else {
+            return original.replace(match, replacement);
+        }
+    }
+    
+    @Override
+    public void onHttpRequestSend(HttpMessage msg, int initiator, HttpSender httpSender) {
+        for (ReplacerParamRule rule : this.getParams().getRules()) {
+            if (rule.isEnabled() && rule.appliesToInitiator(initiator)) {
+                Pattern p = null;
+                if (rule.isMatchRegex()) {
+                    p = Pattern.compile(rule.getMatchString());
+                }
+                switch (rule.getMatchType()) {
+                case req_header:
+                    LOGGER.debug("Add in request header: " + rule.getMatchString() + " : " + rule.getReplacement());
+                    msg.getRequestHeader().setHeader(rule.getMatchString(), rule.getReplacement());

```

Comment:
From the description of the help page, the header shouldn't be being added if the replacement text is empty?